### DISCUSSION TITLE
Attaching callback onto tile loads

### DISCUSSION THREAD

**Discussion by JAremko** [openglobus/openglobus#823] [Q&A] [answered]
I have a big set of "points of interest" that I want to place on the globe lazily.  For this I developed an API that mirrors OSM tile server's one: `XXX.XXX/{z}/{x}/{y}.poi` that returns sets of the markers for the tile specified in the request. 

Now I need a way to hook into the tile display system so that when Openglobus prepares to display a tile I'll be able to translate its coordinates into corresponding request to my POI API.

**Comment by JAremko** [comment]
At this point, the best solution I have found is adding a sniffer to `urlRewrite`:
```  ts
import {XYZ, IXYZParams} from "@openglobus/og";
import {Segment} from "@openglobus/og/lib/js/segment/Segment";
import {stringTemplate} from "@openglobus/og/lib/js/utils/shared";

function toQuadKey(x: number, y: number, z: number) {
    var index = '';
    for (let i = z; i > 0; i--) {
        var b = 0;
        var mask = 1 << (i - 1);
        if ((x & mask) !== 0) b++;
        if ((y & mask) !== 0) b += 2;
        index += b.toString();
    }
    return index;
}

export class SatLayer extends XYZ {
    constructor() {
        const name ="SatLayer"
        super(name, {
            iconSrc: "https://ecn.t0.tiles.virtualearth.net/tiles/a120.jpeg?n=z&g=7146",
            subdomains: ['t0', 't1', 't2', 't3'],
            url: "https://ecn.{s}.tiles.virtualearth.net/tiles/a{quad}.jpeg?n=z&g=7146",
            isBaseLayer: true,
            maxNativeZoom: 17,
            defaultTextures: [{color: "#001522"}, {color: "#E4E6F3"}],
            attribution: `<div style="transform: scale(0.8); margin-top:-2px;"><a href="https://www.bing.com" target="_blank"><img style="position: relative; top: 2px;" title="Bing Imagery" src="https://sandcastle.cesium.com/CesiumUnminified/Assets/Images/bing_maps_credit.png"></a> © 2021 Microsoft Corporation</div>`,
            urlRewrite: (s: Segment, u: string) => {
                console.log(s.tileX, s.tileY, s.tileZoom);
                return stringTemplate(u, {
                    's': this._getSubdomain(),
                    'quad': toQuadKey(s.tileX, s.tileY, s.tileZoom)
                });
            },
            specular: [0.00063, 0.00055, 0.00032],
            ambient: "rgb(90,90,90)",
            diffuse: "rgb(350,350,350)",
            shininess: 20,
            nightTextureCoefficient: 2.7,
        });
    }
}
```

**Comment by Zemledelec** [comment, marked as answer]
Hi @JAremko! It's an interesting architecture, but it looks like you're getting points from the XYZ layer and then converting them to billboards or 3d objects. I use a different approach, I get points in camera radius and elevation and then queue and render the points on the map using the Vector layer as billboards or geo-objects. But this uses a different API instead of XYZ tiles.

In your last solution with satLayer XYZ, globus creates an image layer that is rendered as an image layer, this is not an optimal solution. Another way is to use [CanvasTiles](https://sandbox.openglobus.org/examples/canvasTiles) to get tile data.

Unfortunately, there is no suitable abstract mechanism to retrieve custom tile data.   All we have right now is a list of visible nodes (tails) that we see on each frame, Planet._visibleNodes, which we can read in the planet.on("draw") event, but you will need to create your own algorithm for how to properly handle the loading. For example, in the case of the terrain provider, I wait until the parent terrain is ready.

OK, I suppose your solution isn't too bad (as a workaround), but it's a good reason to think about a better, abstract tile layer in the future. It would certainly be useful for working with vector tile data.

Thanks @JAremko !
